Automattic/jetpack
# Only show one upgrade nudge for nested paid blocks

Fixes 221-gh-Automattic/view-design

When paid blocks are nested inside each other, the UI becomes cluttered with stacked upgrade nudges. For instance, if `plan_check` is enabled on both the Send-a-Message block and its child WhatsApp-Button, you get this in both the editor and the frontend:
<img width="796" alt="Screen Shot 2021-01-20 at 10 47 42 AM" src="https://user-images.githubusercontent.com/63313398/113460141-893cb280-93cc-11eb-84b5-0470424a506e.png">

The purpose of this PR is to prevent multiple stacked nudges from rendering, by not displaying nudges for a block if one of its parents is already displaying the nudge. The goals here are:

* The frontend nudge should not render for a block if one of its parent blocks is _already rendering a nudge_
* The block developer should not have to configure anything extra when creating a new paid block -- ie, they should be able to add `plan_check => true` and have this taken care of for them
* The nudge behavior should be consistent for all paid blocks.
#### Changes proposed in this Pull Request:

* The `WithUpgradeBanner` higher order component takes care of setting `shouldDisplayFrontendBanner` (introduced in #19768) to false if this block is nested within another block with a nudge.
* Adds `plan_check` back to the Send-a-Message block (it was removed in the past to avoid dealing with the duplicate nudges it causes on WhatsApp blocks)

Some other modifications to in-editor upgrade nudges:
* Banner now displays for a block if it is selected or if _any descendent not rendering its own banner_ is selected (previously only worked for immediate children). Example: Premium Content block,  previously if you clicked the text inside the block the banner would disappear; now it stays up while the block is being edited.
* Banners only display for the "most specific" possible selection -- eg, if you have two nested paid blocks and click the outer block, its nudge will display; if you click the inner block, its nudge will display.

**Notes:**
The attribute is set to true by default so that frontend nudges on posts created before this PR is merged will still work, but this means that all possible banners will display on these posts. This is unlikely to be an issue because:

1. Frontend nudges are only visible to admins.
2. If the admin opens the post in the editor and saves, nested nudges will be removed from that point on.
3. Nested nudges are only especially unsightly if the user has done something like nesting a WhatsApp button inside Premium Content, which seems unlikely.

**Known Issues:**
There is a bug with the logic for setting `isChildBannerVisible` when a block has multiple paid-block children that are siblings to one another. Repro:

* Create a Premium Content block and insert a WhatsApp button into the Visitor View, before the LogIn/Subscribe buttons. The WhatsApp button and Recurring Payments (Subscribe button) are siblings to one another and both children of Premium Content.
* Click the Subscribe button. Its banner renders
* Click the WhatsApp button.

Expected: WhatsApp banner renders, and Subscribe/Premium Content banners do not.
Actual: WhatsApp banner and Premium Content banners render, Subscribe does not.

What's Happening:
Because the WhatsApp button is above the Subscribe button, it renders first and calls `setIsChildBannerVisible( true )` on its parent (Premium Content). The Subscribe button renders second and calls `setIsChildBannerVisible( false )` on Premium Content as its banner is no longer displaying. Premium Content believes it has no visible child banner and thus displays its own banner.


#### Jetpack product discussion

* N/A

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:


**Note for testing: the Recurring Payments block is not configured with `plan_check`. Its banner will display in the editor but not on the frontend; that is outside of the scope of this PR.

* Apply the diff on your WordPress.com sandbox.
* Insert a WhatsApp block into a post.
* Verify that only one upgrade nudge displays in the editor. Only one nudge should continue to display if you click up into the `Send-a-Message` block.
* Verify that only one upgrade nudge displays on the frontend
* Insert a Premium Content block into a post.
* Insert a WhatsApp block inside the Subscriber View of the Premium Content block.
* Verify that when you click the WhatsApp block,  only its nudge displays; when you click the Premium Content block, or anything inside the block other than the WhatsApp button, only _its_ nudge displays.
* Verify that you see only the nudge for the Premium Content block on the frontend.
* Stop sandboxing your site, and create a new post containing a Premium Content block. Save and view it on the frontend and you should see the frontend nudge.
* Turn the sandbox back on and view the post in the frontend again _without updating the post_. You should still see the frontend nudge.
* Open the post in the editor and verify there are no console errors.

**See above for known bugs**

#### Screenshots

Editor:
| Before | After |
| --- | --- |
| ![nested-blocks-before](https://user-images.githubusercontent.com/63313398/117377267-b971e800-ae87-11eb-9734-19ff2c5123aa.gif) | ![nested-blocks-working](https://user-images.githubusercontent.com/63313398/117377275-c1318c80-ae87-11eb-8803-a031d1fcc472.gif) |
| _Note that the Premium Content banner disappears when you click inside the block and is difficult to get back_ | |

Frontend:
In each of these pictures, the first WhatsApp block is outside of the Premium Content block, and the second is nested inside.
| Before | After |
| --- | --- |
| <img width="810" alt="Screen Shot 2021-05-06 at 4 29 46 PM" src="https://user-images.githubusercontent.com/63313398/117377562-68aebf00-ae88-11eb-820e-04fe004d75c2.png"> | <img width="776" alt="Screen Shot 2021-05-05 at 12 59 17 PM" src="https://user-images.githubusercontent.com/63313398/117377640-8f6cf580-ae88-11eb-94c4-c31bc3c45efd.png"> |
